### DISCUSSION TITLE
Time to start Optimization for Large Problems

### DISCUSSION THREAD

**Discussion by FreddyHunter** [google/or-tools#3167] [Routing (and legacy CP) questions]
Dear OR-Tools Group,

I am developing a router system using OR-Tools, wrote in python and running at Ubuntu 20.04. My model considers vehicle time windows, location time windows, weight and volume constraints per vehicle, different quantity of stops per vehicle and etc. We have being testing many different size problems, but last week we started to test a potential customer which had 800 orders to router in one day, considering a fleet of around 50 vehicles from 4 types (small, medium, large and extra large in terms of capacity) and just one warehouse, but a single trip (start at warehouse and finishes somewhere else). For our surprise it took more than an hour to start optimizing and close to 4 hours to execute around 400 simulations, as you can see below an extract of our log file. Have you faced this kind of situation? Can this time be considered resonable for this size of problem?

01:45:17.953496  1036 search.cc:264] Root node processed (time = 39 ms, constraints = 18050, memory used = 152.79 MB)
03:19:27.277926  1036 search.cc:264] Solution #0 (3518950873871, time = 5649363 ms, branches = 12533671, failures = 6266475, depth = 33, memory used = 154.68 MB, limit = 31%)
03:19:28.519752  1036 search.cc:264] Solution #1 (3518950751870, objective maximum = 3518950873871, time = 5650605 ms, branches = 12533780, failures = 6266515, depth = 33, Reloca>
03:19:29.968400  1036 search.cc:264] Solution #2 (3518950140370, objective maximum = 3518950873871, time = 5652054 ms, branches = 12533890, failures = 6266564, depth = 33, Reloca>
03:19:31.651943  1036 search.cc:264] Solution #3 (3518949542370, objective maximum = 3518950873871, time = 5653737 ms, branches = 12533999, failures = 6266629, depth = 33, Reloca>
03:19:33.121744  1036 search.cc:264] Solution #4 (3518948992871, objective maximum = 3518950873871, time = 5655207 ms, branches = 12534110, failures = 6266680, depth = 33, Reloca>
03:19:36.166442  1036 search.cc:264] Solution #5 (3518948958868, objective maximum = 3518950873871, time = 5658252 ms, branches = 12534219, failures = 6266793, depth = 33, Reloca>
03:19:36.244078  1036 search.cc:264] Solution #6 (3518948671868, objective maximum = 3518950873871, time = 5658329 ms, branches = 12534329, failures = 6266795, depth = 33, Reloca>
03:19:36.704193  1036 search.cc:264] Solution #7 (3518947256870, objective maximum = 3518950873871, time = 565878

...

06:35:08.642929  1036 search.cc:264] Solution #420 (3496163627369, objective maximum = 3518950873871, time = 17390728 ms, branches = 12576752, failures = 6698216, depth = 33, Swa>
06:35:08.705061  1036 search.cc:264] Solution #421 (3496138547369, objective maximum = 3518950873871, time = 17390790 ms, branches = 12576847, failures = 6698218, depth = 33, Swa>
06:36:27.581209  1036 search.cc:264] Solution #422 (3496111933869, objective maximum = 3518950873871, time = 17469666 ms, branches = 12576943, failures = 6701115, depth = 33, Swa>
06:37:48.259954  1036 search.cc:264] Solution #423 (3496095451369, objective maximum = 3518950873871, time = 17550345 ms, branches = 12577038, failures = 6704096, depth = 33, Swa>
06:39:25.766977  1036 search.cc:264] Solution #424 (3496018584869, objective maximum = 3518950873871, time = 17647852 ms, branches = 12577136, failures = 6707663, depth = 33, Swa>
06:41:02.334474  1036 search.cc:264] Solution #425 (3495924488369, objective maximum = 3518950873871, time = 17744420 ms, branches = 12577231, failures = 6711217, depth = 33, Swa>
06:42:39.087773  1036 search.cc:264] Solution #426 (3495911812869, objective maximum = 3518950873871, time = 17841173 ms, branches = 12577327, failures = 6714776, depth = 33, Swa>
06:42:44.801320  1036 search.cc:264] Solution #427 (3495889382869, objective maximum = 3518950873871, time = 17846886 ms, branches = 12577422, failures = 6714989, depth = 33, Swa>
06:43:04.936662  1036 search.cc:264] Finished search tree (time = 17867022 ms, branches = 12577427, failures = 6715740, neighbors = 6296275, filtered neighbors = 448806, accepted>
06:43:04.944685  1036 search.cc:264] End search (time = 17867029 ms, branches = 12577427, failures = 6715740, memory used = 154.68 MB, speed = 703 branches/s)

We are defining our data using lists, have someone tried develop the code instead of using lists using numpy? Have you got a hugh performance improvement doing that?

Thanks for your help and contribution.

**Comment by lperron** [comment]
Have you enabled local search ?

**Reply by FreddyHunter** [reply]
Hi Laurrent,

thanks for your reply. We have set like the following, is it the best one to be used?

    search_parameters = pywrapcp.DefaultRoutingSearchParameters()
    search_parameters.first_solution_strategy = (
        routing_enums_pb2.FirstSolutionStrategy.PATH_CHEAPEST_ARC)
    search_parameters.local_search_metaheuristic = (
        routing_enums_pb2.LocalSearchMetaheuristic.GUIDED_LOCAL_SEARCH)
    search_parameters.log_search = True
    search_parameters.lns_time_limit.seconds = 1

Thanks for your attention and support.

**Comment by jmarca** [comment]
What language are you using?  If not c++ then I suggest turning on caching if you haven’t done so already. 

James

**Reply by FreddyHunter** [reply]
Hi James,

thanks for your reply. I am using Python version, but I never used caching, do you have an example to share?

Thanks for your attention and support.

**Comment by jmarca** [comment]
Try something like:

```python
model_parameters = pywrapcp.DefaultRoutingModelParameters()
# tell the solver a really large number for cache size.  I used 2x(nodes**2)
model_parameters.max_callback_cache_size = num_nodes * num_nodes * 2
# Create Routing Model with the model parameters
routing = pywrapcp.RoutingModel(self.manager, model_parameters)
```


On Thu, Mar 10, 2022 at 01:29:44PM -0800, FreddyHunter wrote:
> Hi James,
>
> thanks for your reply. I am using Python version, but I never used caching, do you have an example to share?
>
> Thanks for your attention and support.
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/google/or-tools/discussions/3167#discussioncomment-2334851
> You are receiving this because you commented.
>
> Message ID: ***@***.***>

--

James E. Marca
Activimetrics LLC

**Reply by FreddyHunter** [reply]
Hi James,
thanks for your quick return. I will try this, but where I can find all the possible parameters we can set? I also noticed that there is another one called reduce_vehicle_cost_model, do you know or can explain when to use it and how to use?
Thanks.

**Reply by jmarca** [reply]
It is all in the source code...

**Comment by FreddyHunter** [comment]
Hi James, how are you?

thanks for your tip, this adjustment improved a lot the performance. I also
checked the source to understand other options in terms of parameters, but
it was not clear for me about the reduce_vehicle_cost_model. Do you know
how is the impact of it, and how it is activated?

Thanks,
Freddy.

Em qui., 10 de mar. de 2022 às 20:44, James Marca ***@***.***>
escreveu:

**Reply by jmarca** [reply]
Yes, caching the dimension calls is an easy win. I think using the 2d dimension is faster, but more of a pain to set up. 

As to the other parameters, I haven’t needed to use them so I haven’t explored them yet. Some day I might have a project that calls for them, but not yet.

**Reply by jmarca** [reply]
After a quick look at it, it seems that parameter guides whether to simplify the calculation of the vehicle costs. Reduce here means reducing algorithmic complexity. 

If your vehicle costs are the same, this can be true (the default); if not, then I guess it should be false, but that will probably slow down execution, not speed it up.
